### PR TITLE
Feature/dock 1731/discourse check

### PR DESCRIPTION
**Pull request by fhembroff**
This PR adds a build step that ensures a discourse topic is included on all new PR requests. Currently, this check will only run on files modified in a PR, and not on all files.

This is important as a lot of rst files do not contain discourse topics. See this [issue](https://github.com/dockstore/dockstore/issues/4953) for more details.

Unfortunately, this check can only be done on PRs and not on regular pushes. This is because I cannot find a way to get the base branch (ie. master or develop) from a push, only a PR. Given the fact that CircleCI will not run again if a push is turned into a PR, I have created a pull request template that asks the person creating the PR to run CircleCI again to ensure the discourse check is ran. 

**Issue:**
[DOCK-1731](https://ucsc-cgl.atlassian.net/browse/DOCK-1731)
https://github.com/dockstore/dockstore/issues/4081

**Notes:**
The following files,
- `docs/advanced-topics/best-practices/best-practices-secure-fair-workflows.rst`
- `docs/advanced-topics/docker-alternatives.rst`

contained broken links. So I replaced them with links from the web archive. In addition, since these were updated and do not contain discourse topics they were flagged by the new build step. Since I don't have admin privileges in discuss.dockstore I was unable to create discourse topics, so I added them to `no-discourse-topic-required.txt`.

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+**Description**
+A description of this PR.
+
+Please make sure that you've checked the following before submitting your pull request. Thanks!
+
+- [ ] Please rerun the CircleCI tests after creating this PR. This will ensure the job **Discoure Topic Check** is ran correctly
```
Comment (denis-yuen) [inline .github/pull_request_template.md:6]: Discoure typo, also in title of PR

File: no-discourse-topic-required.txt [hunk 2]
```diff
@@ -0,0 +1,4 @@
+docs/index.rst
```
Comment (denis-yuen) [inline no-discourse-topic-required.txt:1]: In one of your DMs, you had a much bigger list (on the order of tens of topics), was that a false alarm or do I misunderstand something?

**Review by denis-yuen — COMMENTED** [review]
Nice! 
Have one question really

File: .github/pull_request_template.md (hunk 1 above)
Comment (fhembroff) [reply .github/pull_request_template.md]: Corrected the typo.

**Review by coverbeck — COMMENTED** [review]
> Given the fact that CircleCI will not run again if a push is turned into a PR,

Is this a per-repo setting? I could swear it does that in the web service repo, but maybe I'm wrong.

> This is because I cannot find a way to get the base branch (ie. master or develop)

The UI has a similar concept for audits (only check for new ones), https://github.com/dockstore/dockstore-ui2/blob/04db396d2eccd0b939a83d5fe8f95b6f5b02fbdc/scripts/npm-audit-comparison.sh#L24, where it defaults to `develop`, from the package.json.

I wonder if it's better to do something like that, or just even hard-code it to develop, rather than relying on engineers remembering to launch the tests.

File: .github/pull_request_template.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+**Description**
+A description of this PR.
+
+Please make sure that you've checked the following before submitting your pull request. Thanks!
+
+- [ ] Please rerun the CircleCI tests after creating this PR. This will ensure the job **Discourse Topic Check** is ran correctly
```
[the remark this answers is not in the document]
Comment (fhembroff) [reply .github/pull_request_template.md]: Corrected

File: discourse-topic-check.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    return 1
+  else
+    return 0
+  fi
+}
+
+RETURN_VALUE=0
+DOES_NOT_REQUIRE_DISCOURSE_TOPIC=no-discourse-topic-required.txt
+
+# Determine the base branch (ie. master or develop) from the PR
```
[the remark this answers is not in the document]
Comment (fhembroff) [reply discourse-topic-check.sh:36]: I have added a comment explaining the use of CIRCLE_PULL_REQUEST.

The point of this variable is to know what branch (master or develop) the PR originated from. This way we are only checking files that were modified for discourse topics.

**Comment by fhembroff** [comment]
> The UI has a similar concept for audits (only check for new ones), https://github.com/dockstore/dockstore-ui2/blob/04db396d2eccd0b939a83d5fe8f95b6f5b02fbdc/scripts/npm-audit-comparison.sh#L24, where it defaults to `develop`, from the package.json.
> 
> I wonder if it's better to do something like that, or just even hard-code it to develop, rather than relying on engineers remembering to launch the tests.

@coverbeck I have made it so it checks to see if it is a PR, and if it is then it runs the check against the base branch of the PR. If it isn't a PR it assumes that the base branch is develop.

**Review by coverbeck — APPROVED** [review]

**Pull request merged**